### PR TITLE
Remove pkwrite dependency in pk using PSA for ECDSA

### PR DESCRIPTION
**Pull request by gilles-peskine-arm**
Remove the dependency on `MBEDTLS_PK_WRITE_C` when `MBEDTLS_PSA_CRYPTO_C` and `MBEDTLS_PK_C` are enabled but not `MBEDTLS_RSA_C`, by no longer using pkwrite functions for ECDSA.

Status: ~needs to be rebased now that https://github.com/Mbed-TLS/mbedtls/issues/6408 has been fixed.~ done
~Depends on #6970~

## Gatekeeper checklist

- [x] **changelog** not required (optimization only)
- [x] **backport** not required (optimization only)
- [X] **tests** provided (covered by existing tests)

### CODE REVIEW

File: include/mbedtls/build_info.h [hunk 1]
```diff
@@ -74,7 +74,11 @@
 #include MBEDTLS_USER_CONFIG_FILE
 #endif
 
-#if defined(MBEDTLS_PK_C) && defined(MBEDTLS_USE_PSA_CRYPTO)
+/* The PK wrappers need pk_write functions to format RSA key objects
+ * when they are dispatching to the PSA API. This happens under USE_PSA_CRYPTO,
+ * and also even without USE_PSA_CRYPTO for mbedtls_pk_sign_ext(). */
```
Comment (gilles-peskine-arm) [inline include/mbedtls/build_info.h:79]: I've now realized that the bit about `mbedtls_pk_sign_ext` is a [regression in 3.2.0](https://github.com/Mbed-TLS/mbedtls/issues/6408). I'm going to make a separate pull request for that, then rebase this one on top of it.

File: library/pk_wrap.c [hunk 2]
```diff
@@ -1164,15 +1116,14 @@ static int ecdsa_sign_wrap( void *ctx_arg, mbedtls_md_type_t md_alg,
     psa_key_attributes_t attributes = PSA_KEY_ATTRIBUTES_INIT;
     mbedtls_svc_key_id_t key_id = MBEDTLS_SVC_KEY_ID_INIT;
     psa_status_t status;
-    mbedtls_pk_context key;
-    size_t key_len;
-    unsigned char buf[MBEDTLS_PK_ECP_PRV_DER_MAX_BYTES];
-    unsigned char *p;
+    unsigned char buf[PSA_KEY_EXPORT_ECC_KEY_PAIR_MAX_SIZE(
+            PSA_VENDOR_ECC_MAX_CURVE_BITS )];
```
Comment (valeriosetti) [inline library/pk_wrap.c:1120]: In case the above comment makes sense to you, what about creating a define to `MBEDTLS_PSA_MAX_EC_PUBKEY_LENGTH` for this purpose?
Perhaps `PSA_EXPORT_PUBLIC_KEY_MAX_SIZE` is already OK for this?

**Review by valeriosetti — COMMENTED** [review]
In general it LGTM. I just have a couple of comments/suggestions for defines

**Comment by gilles-peskine-arm** [comment]
I held off on this while other high-priority work was touching pk. Now I think it would help with the ongoing work on driver-only ECC. But I don't have time for it right now — if someone has time now and wants to adopt it, please feel free. Otherwise I'll get around to it eventually.

**Comment by valeriosetti** [comment]
I just did a force push in order to align with the new code style

File: include/mbedtls/build_info.h (hunk 1 above)
Comment (mpg) [reply include/mbedtls/build_info.h:85]: The regression was fixed and this PR was rebased, so I'm marking this conversation as resolved.

File: library/pk_wrap.c [hunk 3]
```diff
@@ -949,11 +946,12 @@ static int ecdsa_verify_wrap(void *ctx_arg, mbedtls_md_type_t md_alg,
     psa_key_attributes_t attributes = PSA_KEY_ATTRIBUTES_INIT;
     mbedtls_svc_key_id_t key_id = MBEDTLS_SVC_KEY_ID_INIT;
     psa_status_t status;
-    mbedtls_pk_context key;
-    int key_len;
-    unsigned char buf[MBEDTLS_PK_ECP_PUB_DER_MAX_BYTES];
+    size_t key_len;
+    /* This buffer contains first the public key (consisting of two public
+     * points plus a header byte), then the signature (consisting of two
+     * public points). Size it for the public key which is one byte larger. */
```
Comment (mpg) [inline library/pk_wrap.c:952]: Actually, the public key is one point, encoded as one header byte plus two numbers (in the range [0, p)), while the signature is two numbers (in the range [0, n)). For all curves except secp224k1, p and n have the same byte-size, so indeed a public key is one byte larger. Fortunately, we already [don't support secp224k1 with PSA](https://github.com/Mbed-TLS/mbedtls/issues/3541), and are hoping to [remove it entirely in 4.0](https://github.com/Mbed-TLS/mbedtls/issues/6792#issuecomment-1352905076), so I think it's OK to keep using the pubkey length here.

Please just adjust the comment. Suggesting: "The buffer will contain the public key then the signature at different points in time. For all curves except secp224k1, which is not currently supported in PSA, the public key is one byte longer (header byte + 2 numbers, while the signature is only 2 numbers), so use that as the buffer size."

File: tests/suites/test_suite_pk.function [hunk 4]
```diff
@@ -1182,7 +1182,7 @@ exit:
 }
 /* END_CASE */
 
-/* BEGIN_CASE depends_on:MBEDTLS_HAS_ALG_SHA_256_VIA_MD_OR_PSA_BASED_ON_USE_PSA:MBEDTLS_USE_PSA_CRYPTO */
+/* BEGIN_CASE depends_on:MBEDTLS_HAS_ALG_SHA_256_VIA_MD_OR_PSA_BASED_ON_USE_PSA:MBEDTLS_USE_PSA_CRYPTO:MBEDTLS_PK_WRITE_C */
```
Comment (mpg) [inline tests/suites/test_suite_pk.function:1185]: This is surprising, I thought the point of the PR was removing that dependency for ECDSA. So, I think the dependency on `PK_WRITE` should not be for the entire function, but just for the test cases (in the .data file) that use this function with RSA.

**Review by mpg — CHANGES_REQUESTED** [review]
Just a partial review, getting myself familiar with this PR.

File: library/pk_wrap.c [hunk 5]
```diff
@@ -19,46 +19,43 @@
...
+#include "pkwrite.h"
 #endif
 
 #if defined(MBEDTLS_USE_PSA_CRYPTO)
 #include "psa/crypto.h"
 #include "mbedtls/psa_util.h"
-#include "mbedtls/asn1.h"
 #include "hash_info.h"
+
+#if defined(MBEDTLS_ECDSA_C)
```
[the remark this answers is not in the document]
Comment (valeriosetti) [reply library/pk_wrap.c:48]: Correct! I will add the proper label to track this

File: tests/suites/test_suite_pk.function (hunk 4 above)
Comment (valeriosetti) [reply tests/suites/test_suite_pk.function]: I tried to solve this problem in [this commit](https://github.com/Mbed-TLS/mbedtls/pull/6389/commits/e6af138bd5b6cc49ba70c2eb8f3891f0b2f6df14), but now there is a failure in `test_suite_pk` because the function `pk_psa_sign` internally uses `mbedtls_pk_write_pubkey_der` (as wrappers did previously).
However instead of solving this as we did [here](https://github.com/Mbed-TLS/mbedtls/pull/6389/commits/59591db4dc0be0667a123a7b18e0ed8472db2656) I suggest to address your comment in [this issue](https://github.com/Mbed-TLS/mbedtls/issues/6009#issuecomment-1413489731) and fix the problem properly on both sides.
Wdyt?

Comment (mpg) [reply tests/suites/test_suite_pk.function]: Ah, sorry, I missed that. Clearly if the test function itself calls a function from PK-write then the dependency in this place was justified. It's not that the library functions that we're testing depend on this, it's that the test code depends on it. Ok then. Sorry for missing it in the first place.

Comment (valeriosetti) [reply tests/suites/test_suite_pk.function]: What if I address this new problem in the same way that was done in `pk_wrap`?
It should go this way:

- in RSA case, we keep the `pk_write` stuff
- in ECDSA one, we go for ECP+BIGNUM

Wdyt?

Comment (mpg) [reply tests/suites/test_suite_pk.function]: Good idea, let's do this!

**Review by mpg — APPROVED** [review]
LGTM. Let's hope the CI agrees.

**Review by mpg — DISMISSED** [review]

**Comment by valeriosetti** [comment]
Following the suggestion [made previously](https://github.com/Mbed-TLS/mbedtls/pull/6389#discussion_r1094296327) and since #6970 was just merged into `development`, this branch was rebased

**Review by mpg — APPROVED** [review]
LGTM.

I've verified that after the last commit there is no occurrence of `ECDSA_C` left in `pk.c` and `pk_wrap.c` other than (1) those associated with restartable and (2) those that guard `#include "mbedtls/ecdsa.h" `.

**Review by mprse — APPROVED** [review]
LGTM, left one question.

File: library/pk_wrap.c [hunk 6]
```diff
@@ -691,11 +688,14 @@ static int ecdsa_verify_wrap(void *ctx_arg, mbedtls_md_type_t md_alg,
...
-    mbedtls_pk_context key;
-    int key_len;
-    unsigned char buf[MBEDTLS_PK_ECP_PUB_DER_MAX_BYTES];
+    size_t key_len;
+    /* This buffer will initially contain the public key and then the signature
+     * but at different points in time. For all curves except secp224k1, which
+     * is not currently supported in PSA, the public key is one byte longer
+     * (header byte + 2 numbers, while the signature is only 2 numbers),
+     * so use that as the buffer size. */
+    unsigned char buf[MBEDTLS_PSA_MAX_EC_PUBKEY_LENGTH];
```
[the remark this answers is not in the document]
Comment (mpg) [reply library/pk_wrap.c:697]: Ah, excellent question! As the name implies, public keys are not supposed to be sensitive data so they don't need to be wiped out. I'd say signatures are not sensitive either in general, but I may be missing some corner cases, so I'll ask on slack.
